src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverter.cs

```diff
@@ -382,15 +396,39 @@ private static bool TryParseEnumCore(ref Utf8JsonReader reader, JsonSerializerOp
             return success;
         }
 #else
-        private static bool TryParseEnumCore(string? enumString, JsonSerializerOptions _, out T value)
+        private bool TryParseEnumCore(string? enumString, JsonSerializerOptions _, out T value)
         {
+            if ((_converterOptions & EnumConverterOptions.AllowNumbers) == 0 && (enumString is null || !CheckIfValidNamedEnumLiteral(enumString.AsSpan())))
+            {
+                value = default;
+                return false;
+            }
+
             // Try parsing case sensitive first
             bool success = Enum.TryParse(enumString, out T result) || Enum.TryParse(enumString, ignoreCase: true, out result);
             value = result;
             return success;
         }
 #endif
 
+        // By just checking first char, will filter out all numeric value, and other named ones which is invalid for named enum literal
+        private static bool CheckIfValidNamedEnumLiteral(ReadOnlySpan<char> source)
```

Comment:
This might be valid for C# but it's not valid for CIL enums in general, consider this F# counterexample:

https://sharplab.io/#v2:DYLgZgzgNALiCWwA+B7ADgUwHYAIDKAnhDBgLYCwAUFVTAZjgLIECiWArqTgLxU784kOAAbCADCBwA3AIbB4AExzEZAJxg4UYHDNzZOOLDNIZRPHGL4ChogNS6UMABYZV0uYpz6uRk2e44AIxW/DbCAISROlgEzvBYAOY4CSgYEDiR4f44AExAA=

Reply:
If consider all is valid for CIL enums, what is expected behavior for following case if `allowIntegerValues` is false:

[pseudo c# to represent one valid IL]
`   
 enum MyEnum
    {
        1 = 1,
        2 = 2
    }
    // coming json is: "1"
`
**"1"** is integer but is also valid enum named constant in CIL. Should we allow **"1"** or not, in this case ?

Reply:
I think it should be allowed since it matches one of the enum labels.

Reply:
> it should be allowed since it matches one of the enum labels.

I just have a look at behavior of current `Enum.TryParse()` and seems it is different: when input string is `"2"`, `Enum.TryParse()` outputs `MyEnum `value **whose** **underlying integer** value is `2` rather than `-2`, which means it treats `2` as underlying integer, not enum label; 
On `EnumConverter` side, If `"2"` should be allowed, it will output value **whose underlying integer** is `-2` ?

![image](https://user-images.githubusercontent.com/72254037/219000422-86c8408c-2d54-49bb-be6c-1abed4a53cc2.png)
![image](https://user-images.githubusercontent.com/72254037/219000638-f9994308-c61e-4651-92ca-bb3e5da7adef.png)

Output log:
`True, 2`

Reply:
> On EnumConverter side, If "2" should be allowed, it will output value whose underlying integer is -2

I think this is reasonable when we're using `StringEnumConverter`. The underlying numeric value should not be consulted at all in that case.

Reply:
Ok, so in case as above, when use `StringEnumConverter` for deserialization, string-typed json "2" will be deserialized to MyEnum instance called '**myEnum1**' (which = -2) then used somewhere..
Then call Enum.TryParse("2", out MyEnum **myEnum2**), thus **myEnum2** will be 2.
Eventually `myEnum1 == myEnum2` will be `false`.

Just confirm, is that ok?

Reply:
> provided of course that we reject `"-2"` inputs

Yes.

> Yes, I think that should be ok

Because current `EnumConverter` used `Enum.TryParse()` (no matter `allowInteger` is true/false) so both output 2; if we decide to have behavior described above,` EnumConverter` will turn to output -2; if we make true/false `allowInteger` same new behavior, not sure if both have breaking change in this case? Thoughts ?

Reply:
> so both output 2

> will turn to output -2

By output in this case I suppose you mean the deserialization result? How are these different in your example? `2` is a label and `-2` is the numeric representation of the identical value. Could you clarify?

Reply:
> By output in this case I suppose you mean the deserialization result?

Yes, all output are deserialization result and all are underlying numeric value.

So in last example, if input json string `"2"`,
current EnumConverter -> deserialize to MyEnum instance as `2` (it is underlying numeric value, undefined in MyEnum, of course) (it is because current EnumConverter used Enum.TryParse() whose behavior we described as above)
New behavior EnumConverter -> deserialize to MyEnum instance as `-2` (it is also underlying numeric value)
So 2 MyEnum instances are different.

Reply:
Yes this sounds right to me. The explicit goal of https://github.com/dotnet/runtime/issues/58247 is to no longer support deserializing numeric strings as numeric values. I wouldn't be too concerned about breakiness in this particular case as I've never seen real instances of enum labels using numeric values.

Reply:
Got it, based on explicit goal of #58247 and behavior all mentioned in this thread, summary as following rough logic flow, please check the understanding:

**If allowInteger is true:** try to match by enum labels (no matter input is numeric string or not) **firstly**, if fails then try to match by current enum numeric range if input is numeric string. (**Note** this reverts matching priority of `Enum.TryParse` as mentioned so new implementation cannot use `Enum.TryParse` anymore)
**If allowInteger is false:** try to match only by enum labels (again, no matter input is numeric string or not, which means even numeric string is allowed to have a try against enum label), and if label matching fails, deserialization failed.

Reply:
That sounds about right. 

> Note this reverts matching priority of Enum.TryParse as mentioned so new implementation cannot use Enum.TryParse anymore

That sucks, I hope it doesn't make the converter much more complicated than it currently is. If it is, maybe we could come to the conclusion that we just keep the current behavior but it still is worth a shot.

Reply:
Haven't found easy and efficient way to change to new behavior outside `System.Private.CoreLib` (and without `Enum.TryParse`) so still keep current behavior (always treat numeric string value as numeric value rather than enum label) now.
There is just one note: even if input is numeric string value, `EnumConverter` will still try to use `namingPolicy` (if configured) to match it based on background of [this comment](https://github.com/dotnet/runtime/pull/79432#discussion_r1052601700).

Fixed by considering all valid IL identifiers, so not only check first char of input string (only for C#) but check if current whole string value is numeric (implemented by checking on existing utf8 buffer directly to avoid additional utf16-based char array allocation), please review.

Reply:
@eiriktsarpalis any chance to have a look at it ?